src/encode.rs

```diff
@@ -250,7 +250,13 @@ fn encode<'t, A, T>(
                             },
                         }
                     }
-                    pattern.push_str(nsepexpr!("&&{0}]"));
+                    if class.is_negated() {
```

Comment:
This code now repeats the same branch predicate (`class.is_negated()`) with some common code in between. I'd prefer to factor the common code into a function and use only one branch:

```rust
...
use crate::token::Class as ClassToken;

fn encode_class_archetypes(class: &ClassToken, pattern: &mut String) {
    for archetype in class.archetypes() {
        ...
    }
}

let mut pattern = String::new();
pattern.push('[');
if class.is_negated() {
    ...
    encode_class_archetypes(&class, &mut pattern);
    ...
}
else {
    ...
    encode_class_archetypes(&class, &mut pattern);
    ...
}
pattern.push_str("]");
...
```

Reply:
I've refactored this in an additional commit in this PR.